azure-ipam/ipam_test.go

```diff
@@ -0,0 +1,303 @@
+			},
+			Response: cns.Response{
+				ReturnCode: 0,
+				Message:    "",
+			},
+		}
+		return result, nil
+	}
+}
+
+func (c *MockCNSClient) ReleaseIPAddress(ctx context.Context, ipconfig cns.IPConfigRequest) error {
+	switch ipconfig.InfraContainerID {
+	case "failRequestCNSReleaseIPArgs":
+		return errFoo
+	default:
+		return nil
+	}
+}
+
+const (
+	happyPodArgs    = "K8S_POD_NAMESPACE=testns;K8S_POD_NAME=testname;K8S_POD_INFRA_CONTAINER_ID=testid"
+	nothappyPodArgs = "K8S_POD_NAMESPACE=testns;K8S_POD_NAME=testname;K8S_POD_INFRA_CONTAINER_ID=testid;break=break"
+)
+
+type scenario struct {
+	name    string
+	args    *cniSkel.CmdArgs
+	wantErr bool
+}
+
+func TestCmdAdd(t *testing.T) {
+	req := require.New(t)
+
+	happyNetConf := &cniTypes.NetConf{
+		CNIVersion: "0.1.0",
+		Name:       "happynetconf",
+	}
+
+	invalidVersionNetConf := &cniTypes.NetConf{
+		CNIVersion: "0",
+		Name:       "nothappynetconf",
+	}
+
+	happyNetConfByteArr, err := json.Marshal(happyNetConf)
+	if err != nil {
+		panic(err)
+	}
+	invalidVersionNetConfByteArr, err := json.Marshal(invalidVersionNetConf)
+	if err != nil {
+		panic(err)
+	}
+	invalidNetConf := []byte("invalidNetConf")
+
+	happyArgs := &cniSkel.CmdArgs{
+		ContainerID: "happyArgs",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   happyNetConfByteArr,
+	}
+
+	failCreateCNSReqArgs := &cniSkel.CmdArgs{
+		ContainerID: "failCreateCNSReqArgs",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        nothappyPodArgs,
+		StdinData:   happyNetConfByteArr,
+	}
+
+	failRequestCNSArgs := &cniSkel.CmdArgs{
+		ContainerID: "failRequestCNSArgs",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   happyNetConfByteArr,
+	}
+
+	failProcessCNSResp := &cniSkel.CmdArgs{
+		ContainerID: "failProcessCNSResp",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   happyNetConfByteArr,
+	}
+
+	failParseNetConf := &cniSkel.CmdArgs{
+		ContainerID: "failParseNetConf",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   invalidNetConf,
+	}
+
+	failGetVersionedResult := &cniSkel.CmdArgs{
+		ContainerID: "failGetVersionedResult",
+		Netns:       "testnetns",
+		IfName:      "testifname",
+		Args:        happyPodArgs,
+		StdinData:   invalidVersionNetConfByteArr,
+	}
+
+	tests := []scenario{
+		{
+			name:    "Happy CNI add",
+			args:    happyArgs,
+			wantErr: false,
+		},
+		{
+			name:    "Fail create CNS request during CmdAdd",
+			args:    failCreateCNSReqArgs,
+			wantErr: true,
+		},
+		{
+			name:    "Fail request CNS ipconfig during CmdAdd",
+			args:    failRequestCNSArgs,
+			wantErr: true,
+		},
+		{
+			name:    "Fail process CNS response during CmdAdd",
+			args:    failProcessCNSResp,
+			wantErr: true,
+		},
+		{
+			name:    "Fail parse netconf during CmdAdd",
+			args:    failParseNetConf,
```

Comment:
nit: I'm finding this a bit hard to read, because the construction of these args happens in different parts of the file. Consider changing this to a function that does all the construction in one place?

Imagining something like:
```
buildArgs: func() *cniSkel.CmdArgs {
   // do all the setup here...
}
```